hal/src/sercom/v2/uart/charsize.rs

```diff
@@ -0,0 +1,114 @@
+//! Character size definitions
+
+use crate::typelevel::Sealed;
+use num_traits::{AsPrimitive, PrimInt};
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub type DataSize = u16;
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "min-samd51g"))]
+pub type DataSize = u32;
+
+/// Type-level `enum` representing the UART character size
+///
+/// The UART character size affects the word size for the embedded HAL traits.
+/// Eight or less bit transactions use a `u8` word, while nine-bit transactions
+/// use a `u16` word.
+pub trait CharSize: Sealed {
+    /// Word size for the character size
+    type Word: 'static + PrimInt + AsPrimitive<DataSize>;
+
+    /// Bits to write into the `LENGTH` register
+    const BITS: u8;
+}
+
+/// Type-level `enum` indicating a [`CharSize`] that is not dynamic
+pub trait FixedCharSize: CharSize {}
+
+/// Type alias to recover the `Word` type from an implementation of [`CharSize`]
+pub type Word<C> = <C as CharSize>::Word;
+
+/// [`CharSize`] variant for 5-bit transactions
+pub enum FiveBit {}
+
+/// [`CharSize`] variant for 6-bit transactions
+pub enum SixBit {}
+
+/// [`CharSize`] variant for 7-bit transactions
+pub enum SevenBit {}
+
+/// [`CharSize`] variant for 8-bit transactions
+pub enum EightBit {}
+
+/// [`CharSize`] variant for 9-bit transactions
+pub enum NineBit {}
+
+/// Dynamic [`CharSize`] that can be changed on the fly
+pub enum DynCharSize {}
+
+/// `enum` version of [`CharSize`]
+#[repr(u8)]
+pub enum CharSizeEnum {
+    FiveBit = FiveBit::BITS,
+    SixBit = SixBit::BITS,
+    SevenBit = SevenBit::BITS,
+    EightBit = EightBit::BITS,
+    NineBit = NineBit::BITS,
+}
```

Comment:
I think this is a bit roundabout. You define the bits on the type, then use the type to define the enum, then use the enum to get back the bits. I would switch things around.

What if you defined the variants of `CharSizeEnum` directly? Then then `FixedCharSize` trait can define a `const CHAR_SIZE: CharSizeEnum;`. That would also let you modify the corresponding functions in the `reg` module to take a `CharSizeEnum` rather than raw bits. It's a bit dangerous right now, because the `reg` function does not check the bits at all.